[2016.11.0rc1] (re)introduce minimal weighing to get_fqhostname

sjorge opened the pull request:
### What does this PR do?

Introduces minimal weighing to get_fqhostname,
Solaris-like platforms seem to have the nodename as first entry in ai_canonname.
### What issues does this PR fix or reference?
saltstack/salt#36869
### Previous Behavior

Pick first entry
### New Behavior
1. strips out useless '' entries
2. applies minimal weighing to the list

This is based on the old _sort_hostname function, keeping only what made sense in this usecase.
### Tests written?

No

sjorge committed:
(re)introduce minimal weighing to get_fqhostname

sjorge commented:
Looks like unrelated test failures

thatch45 commented:
@isbm is this related to the changes you made in minion name resolution at all?

isbm commented:
@thatch45 yes, and why this even came back? 

@sjorge this weighting reintroducing isn't right. Because: 

1. In Solaris the [nodename is _not_ a hostname](https://docs.oracle.com/cd/E36784_01/html/E36882/nodename-4.html) but a _system_ name, _regardless_ how many actual _hostnames_ it does have. And then it is used in FQDN which calls `socket.getfqdn()`? 
2. Even if this makes any sense at all, it should still stay within Solaris boundaries, so you have to check if this is Solaris or not, excluding systems that does not have Solaris-only things.

thatch45 commented:
I think I would agree on the solaris side of things, we should check for solaris as this does change how the mininon id gets allocated. I will revert this for now, sorry I should have asked before merging

isbm commented:
@sjorge on a total off-topic,  how the heck you got Salt Carbon running on Slowlaris 11?? I am all listening to the instructions, because I wanted to fix many things there too!

sjorge commented:
Seems to be a Solaris specific thing socket.getfqdn() on my hosts (plus some systems I got my hands on at work) seem to retune a null entry + short non fqfn first.

I will resubmit with if Solaris around it.

thatch45 commented:
@isbm I will contact you off channel about solaris and Salt

sjorge commented:
@isbm I mostly test on Solaris 10 when I can at work.

I develop on OmniOS/SmartOS both illumos derivatives. I had feedback that it works on OpenIndiana too, although I do not test there.

isbm commented:
@sjorge This still doesn't look right. The It should rather reuse machine ID. I never got successful even booting SmartOS on KVM instance (no real hardware around). OmniOS I tried once. Maybe you can give me a short info how you get Salt running there? I would help you in this issue _then_.

@thatch45 I spent many years on Solaris and then OpenSolaris before Snorcle destroyed Sun Microsystems and earlier times my name is around in OpenIndiana community. So I know this OS pretty well. Except I stopped looking at it and IPS is a horrid package manager. :smile:

sjorge commented:
@isbm, are you ok with me putting this change in an if block? Without it I get incorrect results on most of my config having to do with hostname/fqdn 

Below are some results form a SmartOS and Solaris 10 system of socket.getfqdn()

```Values of l before sorting
['cronos', 'cronos.vlan30.acheron.be', '']
['tokyo', 'tokyo.private.ugent.be', '']
```

thatch45 commented:
Yes, it is a chore for us to maintain Solaris support, but there are a lot of merits to the Solaris kernel and platform, even though they would be far better off with rpm and zypper :)

sjorge commented:
@isbm I'll write something up when I get home

I'll also dig a bit deeper into this.

sjorge commented:
Also it's only the fqdn grain that was wrong ( other grains using it )

Node, Domain, are all fine.

isbm commented:
OK, I am in. But _somebody_ please help me installing Salt with a minimal pain, so I can invest some time-a-day to clean it up there.

@sjorge This seems a bug in a Python likely. If you can wait (means, maintain your own patch outside the upstream, like [we do](https://build.opensuse.org/package/view_file/systemsmanagement:saltstack:products/salt/0008-Fix-types-in-the-output-data-and-return-just-a-list-.patch?expand=1) sometimes in [our package](https://build.opensuse.org/package/show/systemsmanagement:saltstack:products/salt)) then let get this right, so it won't put @thatch45 again into an `#IDONTKNOW` mode, when asking "What is the purpose of all of this weirdness?" :smile:

sjorge commented:
I'd prefer not to have patches outside of upstream, I've been working really hard to avoid it actually.

I'll dig deeper there must be a solution somewhere that works for all of us.

The easiest route to get a Solaris-like OS running probably still SmartOS. You mentioned you are using QEMU so here is a summary of my notes.

1. grab the smartos bootable iso (https://wiki.smartos.org/display/DOC/Download+SmartOS)
2. add 1 or more nics using the virtio or e1000 device model
3. add 1 disk, IDE (vioblk might also work... but not tested as this driver got upstreamed after my notes)
4. -smp 1 (multiple CPU seem to make it not happy)
5. always boot from the iso
6. you should get to the smartos setup, just go through it setting up one disk.

Once you have SmartOS booted you can create a zone and use ```pkgin``` to install all dependencies and then install salt using pip.

Alternatively you can install the tools pkgsrc bootstrap in the global zone form https://pkgsrc.joyent.com/install-on-illumos/ (click the 64-bit gz-tools tab)

Then use pip to install salt in the gz or use my experimental package from https://pkg.blackdot.be/extras/salt-2016.11rc/salt-2016.11.0rc1_2016Q2_TOOLS.tgz

There is also a esky package for the globalzone but that not really great for debugging: https://pkg.blackdot.be/extras/salt-2016.11rc/salt-2016.11.0rc1-esky-smartos.tar.gz

sjorge commented:
## Test data (unpatched)
### empty etc hosts
/etc/hosts
```
127.0.0.1               localhost localhost.local loghost
::1                     localhost localhost.local loghost
```

/etc/resolv.conf
```
domain acheron.be
search acheron.be mgmt.acheron.be
nameserver 172.16.10.1
nameserver 2001:470:7ee7:10::1
```

/etc/nodename
```
isotope
```

/etc/defaultdomain
```
acheron.be
```

```
[root@isotope /var/cache/salt/minion/extmods]# python
Python 2.7.12 (default, Sep 26 2016, 19:40:04)
[GCC 4.9.3] on sunos5
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.getfqdn()
'isotope'
```

sjorge commented:
(For completion of the date, I removed the dual-homing I added for a test and redeployed the some, this is after a state.apply with a normal etc hosts file)

grain output without patch

```yaml
local:
    ----------
    cpu_flags:
        - rdrand
        - f16c
        - vmx
        - avx
        - xsave
        - pclmulqdq
        - aes
        - sse4.2
        - sse4.1
        - ssse3
        - popcnt
        - tscp
        - cx16
        - sse3
        - sse2
        - sse
        - fxsr
        - mmx
        - cmov
        - amd_sysc
        - cx8
        - tsc
        - fpu
    cpu_model:
        Intel(r) Xeon(r) CPU E5-2620 v2 @ 2.10GHz
    cpuarch:
        amd64
    datacenter:
        Unknown
    dns:
        ----------
        domain:
            acheron.be
        ip4_nameservers:
            - 172.16.10.1
        ip6_nameservers:
            - 2001:470:7ee7:10::1
        nameservers:
            - 172.16.10.1
            - 2001:470:7ee7:10::1
        options:
        search:
            - acheron.be
            - mgmt.acheron.be
        sortlist:
    domain:
    fqdn:
        isotope
    fqdn_ip4:
        - 172.16.10.135
    fqdn_ip6:
        - 2001:470:7ee7:10::135
    gid:
        0
    gpus:
    groupname:
        root
    host:
        isotope
    hwaddr_interfaces:
        ----------
        net0:
            00:22:06:07:10:01
    hypervisor_uuid:
        00000000-0000-0000-0000-002590f15506
    id:
        isotope
    imageversion:
        pkgbuild 15.4.1
    ip4_interfaces:
        ----------
        lo0:
            - 127.0.0.1
        net0:
            - 172.16.10.135
    ip6_interfaces:
        ----------
        lo0:
            - ::1
        net0:
            - fe80::222:6ff:fe07:1001
            - 2001:470:7ee7:10::135
    ip_interfaces:
        ----------
        lo0:
            - 127.0.0.1
            - ::1
        net0:
            - 172.16.10.135
            - fe80::222:6ff:fe07:1001
            - 2001:470:7ee7:10::135
    ipv4:
        - 127.0.0.1
        - 172.16.10.135
    ipv6:
        - ::1
        - 2001:470:7ee7:10::135
        - fe80::222:6ff:fe07:1001
    kernel:
        SunOS
    kernelrelease:
        5.11
    locale_info:
        ----------
        defaultencoding:
            unknown
        defaultlanguage:
            unknown
        detectedencoding:
            646
    localhost:
        isotope
    master:
        cronos.acheron.be
    mem_total:
        16384
    nodename:
        isotope
    num_cpus:
        24
    num_gpus:
        0
    os:
        SmartOS
    os_family:
        Solaris
    osarch:
        amd64
    osfinger:
        SmartOS-20161013T025521Z
    osfullname:
        SmartOS
    osmajorrelease:
        20161013T025521Z
    osrelease:
        20161013T025521Z
    osrelease_info:
        - 20161013T025521Z
    path:
        /usr/local/sbin:/usr/local/bin:/opt/local/sbin:/opt/local/bin:/usr/sbin:/usr/bin:/sbin:/usr/gnu/bin:/usr/sfw/bin:/bin
    pid:
        12136
    pkgsrcpath:
        http://pkgsrc.joyent.com/packages/SmartOS/2015Q4/x86_64/All
    pkgsrcversion:
        2015Q4
    ps:
        /usr/ucb/ps auxwww
    pythonexecutable:
        /opt/local/bin/python2.7
    pythonpath:
        - /var/cache/salt/minion/extmods/utils
        - /opt/local/bin
        - /opt/local/lib/python2.7/site-packages/wheel-0.29.0-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/libnacl-1.5.0-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/ioflo-1.6.1-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/raet-0.6.5-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/smmap-0.9.0-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/gitdb-0.6.4-py2.7-solaris-2.11-i86pc.64bit.egg
        - /opt/local/lib/python2.7/site-packages/python_gnupg-0.3.9-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/psutil-4.3.1-py2.7-solaris-2.11-i86pc.64bit.egg
        - /opt/local/lib/python2.7/site-packages/gnupg-2.0.2-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/timelib-0.2.4-py2.7-solaris-2.11-i86pc.64bit.egg
        - /opt/local/lib/python2.7/site-packages/futures-3.0.5-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/GitPython-0.3.2rc1-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/argparse-1.4.0-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/pytz-2016.6.1-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/dateutils-0.6.6-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/croniter-0.3.12-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/Mako-1.0.4-py2.7.egg
        - /opt/local/lib/python2.7/site-packages/yappi-0.94-py2.7-solaris-2.11-i86pc.64bit.egg
        - /opt/local/lib/python2.7/site-packages/pyzmq-15.4.0-py2.7-solaris-2.11-i86pc.64bit.egg
        - /opt/local/lib/python27.zip
        - /opt/local/lib/python2.7
        - /opt/local/lib/python2.7/plat-sunos5
        - /opt/local/lib/python2.7/lib-tk
        - /opt/local/lib/python2.7/lib-old
        - /opt/local/lib/python2.7/lib-dynload
        - /opt/local/lib/python2.7/site-packages
    pythonversion:
        - 2
        - 7
        - 12
        - final
        - 0
    saltpath:
        /var/cache/salt/minion/extmods
    saltversion:
        2016.11.0rc1
    saltversioninfo:
        - 2016
        - 11
        - 0
        - 0
    server_id:
        1635052238
    shell:
        /opt/local/bin/zsh
    uid:
        0
    username:
        root
    virtual:
        zone
    zmqversion:
        4.1.3
    zoneid:
        16
    zonename:
        b84e8af8-2a3a-e21e-ea6e-83e9d052ab11
```

sjorge commented:
A very minimal patch that should go upstream (doesn't fix the issue)
```patch
--- network.py~ 2016-10-14 19:59:10.531810191 +0200
+++ network.py  2016-10-14 20:02:17.930496949 +0200
@@ -175,7 +175,8 @@
         )
         for info in addrinfo:
             # info struct [family, socktype, proto, canonname, sockaddr]
-            if len(info) >= 4:
+            # canonname sometimes is '', we filter those out
+            if len(info) >= 4 and len(info[3]) > 0:
                 l.append(info[3])
     except socket.gaierror:
         pass
```

This just prevents the '' entries from showing up. Probably not bad to just filter these!

isbm commented:
I would instead filter empty things at the very end in oneliner:
```python
return filter(None, all_my_collected_and_sorted_out_data)
```

isbm commented:
@sjorge but if there are so much quirks, I would rather not hurry up fix Salt, but understand the specific and only then do a patches. Until that I would like to keep experimental stuff away.

Let me try install Salt on this beast and get back to you. BTW, what kind of dependencies before `pip`? You've got a list of them?

sjorge commented:
@isbm ack, but 2016.3 is working pretty well on Solaris-like OS' (OmniOS/SmartOS/Solaris 10/...) So I'd rather not see 2016.11 case some havoc on something like this.

But if there is a better fix, thats even better of course (within salt that is)
Grabbing these via pkgin before pip install salt=='2016.11.0rc0' or 2016.3.3 should be OK:
pkgin in py27-zmq py27-m2crypto py27-crypto py27-msgpack py27-yaml py27-jinja2 py27-apache-libcloud py27-requests py27-tornado

Do note there is no py27-pip package for the global zone, it's only available in a native zone.

isbm commented:
Who needs zones if you have KVM snapshots... :wink:

thatch45 commented:
ok, lets keep it a clean fight ;)

sjorge commented:
@isbm you forgot live migration, SmartOS aint got that :p

sjorge commented:
@isbm heard back from some of the people in illumos! It's a configuration bug on my end!
Sadly enough, the file is wrong on all of works remaining Solaris 10 boxes as well.

The problem was that /etc/nodename should hold the prefered FQDN not just the node part!
Also they noted that when using dhcp the file is ignored, but that did not apply to my setup.

However if you still want to get SmartOS/OmniOS running and get stuck, I am glad to help further with that.

isbm commented:
Well, just like I said before: `/etc/nodename` is _not_ a hostname. It is just an alias and should be considered only when nothing else works. Hence this is why I blew away all that weighting stuff before, because seems like it was written without real understanding how the whole thing supposed to work.

Glad you figured it out.

Yes, I am interested. :laughing:

sjorge commented:
Right, let me know when you run into issues with the stuff I commented earlier in this thread.

As long as /etc/nodename is a FQDN we're good. Same for DHCP with a proper reservation.
